core/Enclave/enclave_hsm.cpp

```diff
@@ -940,4 +940,73 @@ sgx_status_t enclave_verify_quote_policy(uint8_t *quote, uint32_t quote_size,
         return SGX_ERROR_UNEXPECTED;
     }
     return SGX_SUCCESS;
+}
+
+/**
+ * @brief Generate HMAC with given cmk, apikey and payload
+ * @param cmk the cmk for apikey decryption
+ * @param apikey the encrypted apikey
+ * @param payload the payload of HMAC
+ * @param hmac the output of the function
+*/
+sgx_status_t enclave_generate_hmac(ehsm_keyblob_t *cmk, uint32_t cmk_size,
+                                    ehsm_data_t *apikey, uint32_t apikey_size,
+                                    ehsm_data_t *payload, uint32_t payload_size,
+                                    ehsm_data_t *hmac, uint32_t hmac_size)
+{
+    if (cmk == NULL ||
+        cmk_size != APPEND_SIZE_TO_KEYBLOB_T(cmk->keybloblen) ||
+        cmk->keybloblen == 0 ||
+        cmk->metadata.origin != EH_INTERNAL_KEY ||
+        cmk->metadata.keyusage != EH_KEYUSAGE_ENCRYPT_DECRYPT)
+        return SGX_ERROR_INVALID_PARAMETER;
+
+    if (apikey == NULL ||
+        apikey_size != APPEND_SIZE_TO_DATA_T(apikey->datalen))
+        return SGX_ERROR_INVALID_PARAMETER;
+
+    if (payload == NULL ||
+        payload_size != APPEND_SIZE_TO_DATA_T(payload->datalen) ||
+        payload->datalen == 0)
+        return SGX_ERROR_INVALID_PARAMETER;
+
+    if (hmac == NULL ||
+        hmac_size != APPEND_SIZE_TO_DATA_T(hmac->datalen) ||
+        hmac->datalen != EH_HMAC_SHA256_SIZE)
+        return SGX_ERROR_INVALID_PARAMETER;
+
+    sgx_status_t ret = SGX_ERROR_UNEXPECTED;
+    ehsm_data_t *aad = NULL;
+    ehsm_data_t *rawApiKey = NULL;
+
+    // create empty aad
+    aad = (ehsm_data_t *)malloc(APPEND_SIZE_TO_DATA_T(0));
+    if (aad == NULL)
+    {
+        return SGX_ERROR_UNEXPECTED;
+    }
+    aad->datalen = 0;
```

Comment:
even if you don't want to set aad, still need to malloc it?  seems the existing check logic of aad has some issue in current codes?

Reply:
Thanks for reviewing!

The `enclave_decrypt` function does allow passing `aad = NULL` with `aad_size = 0`. However, all tests related to `CreateKey` were failed when passing `aad = NULL` to the `enclave_decrypt` function. I am still investigating the problem.

Reply:
I have traced the call chain of `enclave_decrypt` funciton and found the bug.

The `ehsm_aes_gcm_decrypt` treats `aad` as a non-null variable, directly uses `aad->data` without checking if `aad` is `NULL`.

### Investigation steps

The `enclave_generate_hmac` returns `SGX_ERROR_ENCLAVE_CRASH (0x1006)` when passing `aad=NULL` to `enclave_decrypt` in `enclave_generate_hmac` function.

This is because the `cmk` use `EH_AES_GCM_192` as keyspec, which transfer the control flow to `ehsm_aes_gcm_decrypt()`. However, the `ehsm_aes_gcm_decrypt()` does not check the `aad` varible before the following usage:

```cpp
// https://github.com/intel/ehsm/blob/main/core/Enclave/key_operation.cpp#L390
    ret = aes_gcm_decrypt(key,
                          plaintext->data,
                          block_mode,
                          cipherblob->data,
                          cipherblob->datalen - EH_AES_GCM_IV_SIZE - EH_AES_GCM_MAC_SIZE,
                          aad->data,  // BUG: DEREFERENCE OF NULL
                          aad->datalen, // BUG: DEREFERENCE OF NULL
                          iv,
                          SGX_AESGCM_IV_SIZE,
                          l_tag,
                          SGX_AESGCM_MAC_SIZE);
``` 

Then the enclave is crashed.

### Possible fix

A simple but work fix is changing the call to `aes_gcm_decrypt` in `ehsm_aes_gcm_decrypt` to the following:

```cpp
    ret = aes_gcm_decrypt(key,
                          plaintext->data,
                          block_mode,
                          cipherblob->data,
                          cipherblob->datalen - EH_AES_GCM_IV_SIZE - EH_AES_GCM_MAC_SIZE,
                          aad ? aad->data : NULL,
                          aad ? aad->datalen : NULL,
                          iv,
                          SGX_AESGCM_IV_SIZE,
                          l_tag,
                          SGX_AESGCM_MAC_SIZE);
```

Tests passed after the modification above.

Shall I open a new issue / PR for fixing this?

Reply:
Thanks for the root cause. Of course, please create a separate PR to fix it.  Thanks.

Reply:
yes, agree

Reply:
The new PR #302 is created.

Reply:
The fix patch of #302 has been applied to current branch in the latest commit.

More detailly, `NULL` is passed as `aad` and `0` is passed as `aad_size` when calling `enclave_decrypt` function in `enclave_generate_hmac` function instead of allocating `sizeof(ehsm_data_t)` for `aad` without storing any data.